src/main/java/com/sap/oss/phosphor/fosstars/tool/AbstractHandler.java

```diff
@@ -0,0 +1,472 @@
+   */
+  GitHubDataFetcher fetcher;
+
+  /**
+   * A user callback.
+   */
+  UserCallback callback = NoUserCallback.INSTANCE;
+
+  /**
+   * A cache of values for subjects.
+   */
+  SubjectValueCache cache = new SubjectValueCache();
+
+  /**
+   * A base directory.
+   */
+  String baseDirectory = "./";
+
+  /**
+   * Create a handler.
+   *
+   * @param rating A rating that the handler calculates.
+   */
+  AbstractHandler(Rating rating) {
+    this.rating = requireNonNull(rating, "Oops! Rating is null!");
+    this.router.put("--url", this::processUrl);
+    this.router.put("--gav", this::processMaven);
+    this.router.put("--npm", this::processNpm);
+    this.router.put("--purl", this::processPurl);
+    this.router.put("--config", this::processConfig);
+  }
+
+  /**
+   * Creates a rating calculator that calculates a rating for a single subject.
+   *
+   * @throws IOException If something went wrong.
+   */
+  SingleRatingCalculator calculator() throws IOException {
+    List<DataProvider> providers = dataProviderSelector().providersFor(rating);
+    return new SingleRatingCalculator(rating, providers).set(cache).set(callback);
+  }
+
+  /**
+   * Creates a selector for data providers.
+   *
+   * @return A selector for data providers.
+   * @throws IOException If something went wrong.
+   */
+  DataProviderSelector dataProviderSelector() throws IOException {
+    requireNonNull(fetcher, "Oops! Fetcher is not set!");
+    DataProviderSelector selector = new DataProviderSelector(fetcher, nvd);
+    selector.configure(withDataProviderConfigs);
+    return selector;
+  }
+
+  @Override
+  public List<Option> options() {
+    return emptyList();
+  }
+
+  @Override
+  public List<OptionGroup> optionGroups() {
+    return emptyList();
+  }
+
+  @Override
+  public Handler configureDataProviders(List<String> withDataProviderConfigs) {
+    requireNonNull(withDataProviderConfigs, "Oops! Data provider configs are null!");
+    this.withDataProviderConfigs = new ArrayList<>(withDataProviderConfigs);
+    return this;
+  }
+
+  @Override
+  public Handler baseDirectory(String path) {
+    requireNonNull(path, "Oops! Path is null!");
+    this.baseDirectory = path;
+    return this;
+  }
+
+  @Override
+  public AbstractHandler set(CommandLine commandLine) {
+    requireNonNull(commandLine, "Oops! Command-line options are null!");
+    this.commandLine = commandLine;
+    return this;
+  }
+
+  @Override
+  public Handler set(UserCallback callback) {
+    requireNonNull(callback, "Oops! Callback is null!");
+    this.callback = callback;
+    return this;
+  }
+
+  @Override
+  public Handler set(GitHubDataFetcher fetcher) {
+    requireNonNull(fetcher, "Oops! Fetcher is null!");
+    this.fetcher = fetcher;
+    return this;
+  }
+
+  @Override
+  public Handler set(SubjectValueCache cache) {
+    requireNonNull(cache, "Oops! Cache is null!");
+    this.cache = cache;
+    return this;
+  }
+
+  @Override
+  public void close() {
+
```

Comment:
Could we add some exception here?

Reply:
I was going to use `AutoCloseable` to do a little cleanup in the end, but it is not actually used. I'll just drop it and remove this method.